### ISSUE TITLE
Design: require complete match case coverage?

### ISSUE THREAD

**Issue by wilberton** [nickmqb/muon#10]
The following compiles correctly but leads to a run time crash when mem is neither a Namespace nor a FunctionDef. 
Adding a default block fixes it, but I'm guessing this should really be a compile error.

```
dumpFunctions(ns Namespace) {
	for e in ns.members {
		mem := e.value
		match mem {
			Namespace: {
				dumpFunctions(mem)
			}
			FunctionDef: {
				if (mem.flags & FunctionFlags.foreign) == 0 {
					Stdout.writeLine(format("{}.{}", mem.ns.name, mem.name.value))
				}
			}
		}
	}
}
```

**Comment by nickmqb** [comment]
This is actually a feature: if there are no matching clauses, match will abandon the program. (Whether it's a good feature is perhaps debatable ;), though I think it is). Further down the road, it may be cool to do some static analysis and figure out whether we can determine at compile time whether a match will fail, in that case we can generate a compile error. Though in most cases (like the above) that will probably not be possible to determine at compile time.

**Comment by wilberton** [comment]
Ok, I can kind of see the reasoning behind that. Static analysis in the compiler would be helpful - just out of interest, if you're matching against a discriminated union (like in the above case), why do you say it wouldn't be possible?
For cases when the compiler can't determine ahead of time that all cases are covered, I would much prefer a compile error to a runtime one (especially one with no source/line info about what caused it to abandon).

**Comment by nickmqb** [comment]
When I designed this feature, I considered 3 alternatives w.r.t. to handling unmatched values:
1. Skip the match statement (e.g. this is how switch works in C-like languages)
2. Abandon the program
3. Force complete coverage at compile-time: require explicitly listing all cases and/or default/null cases (which would indeed be possible, since we know all possible options from the tagged pointer def).

I chose option 2, which essentially emits an implicit `default | null: abandon()` clause, so it saves some typing/'noise'. But there's something to be said for option 3 though (and I actually fixed a compiler bug a few weeks ago that would have been prevented by this).
